src/robot_model_builder.cpp

```diff
@@ -0,0 +1,173 @@
+      pinocchio_model_complete_, locked_joint_ids_, q_default_complete_);
+  pinocchio_data_reduced_ = pinocchio::Data(pinocchio_model_reduced_);
+
+  return true;
+}
+
+bool RobotModelBuilder::parseMovingJointNames(
+    const std::vector<std::string>& in_moving_joint_names,
+    const std::vector<std::string>& in_controlled_joint_names,
+    std::vector<std::string>& moving_joint_names,
+    std::vector<pinocchio::Index>& moving_joint_ids,
+    std::vector<pinocchio::Index>& locked_joint_ids) {
+  // Get moving joints ids
+  moving_joint_ids.clear();
+  for (std::vector<std::string>::const_iterator it =
+           in_moving_joint_names.begin();
+       it != in_moving_joint_names.end(); ++it) {
+    const std::string& joint_name = *it;
+    pinocchio::JointIndex joint_id = 0;
+    // do not consider joint that are not in the model
+    if (pinocchio_model_complete_.existJointName(joint_name)) {
+      joint_id = pinocchio_model_complete_.getJointId(joint_name);
+      moving_joint_ids.push_back(joint_id);
+    } else {
+      std::cerr << "joint_name=" << joint_name
+                << " does not belong to the model" << std::endl;
+    }
+  }
+  if (in_robot_has_free_flyer_) {
+    moving_joint_ids.push_back(
+        pinocchio_model_complete_.getJointId("root_joint"));
+  }
+  // Sort them to the pinocchio order (increasing number) and remove duplicates.
+  std::sort(moving_joint_ids.begin(), moving_joint_ids.end());
+  moving_joint_ids.erase(
+      unique(moving_joint_ids.begin(), moving_joint_ids.end()),
+      moving_joint_ids.end());
+
+  // Remap the moving joint names to the Pinocchio order.
+  moving_joint_names.clear();
+  for (std::size_t i = 0; i < moving_joint_ids.size(); ++i) {
+    moving_joint_names.push_back(
+        pinocchio_model_complete_.names[moving_joint_ids[i]]);
+  }
+
+  // Locked joint ids in the Pinocchio order.
+  locked_joint_ids.clear();
+  for (std::vector<std::string>::const_iterator it =
+           pinocchio_model_complete_.names.begin() + 1;
+       it != pinocchio_model_complete_.names.end(); ++it) {
+    const std::string& joint_name = *it;
+    if (std::find(moving_joint_names.begin(), moving_joint_names.end(),
+                  joint_name) == moving_joint_names.end()) {
+      locked_joint_ids.push_back(
+          pinocchio_model_complete_.getJointId(joint_name));
+    }
+  }
+
+  // remove the "root_joint" if it exists in the moving_joint_names, we needed
+  // it to create the above locked_joint_ids needed for the reduced model.
+  auto root_name = std::remove(moving_joint_names.begin(),
+                               moving_joint_names.end(), "root_joint");
+  if (root_name != moving_joint_names.end()) {
+    moving_joint_names.erase(root_name, moving_joint_names.end());
+    auto root_id =
+        std::remove(moving_joint_ids.begin(), moving_joint_ids.end(),
+                    pinocchio_model_complete_.getJointId("root_joint"));
+    moving_joint_ids.erase(root_id, moving_joint_ids.end());
```

Comment:
```suggestion
    const pinocchio::JointIndex root_joint_idx =
        pinocchio_model_complete_.getJointId("root_joint");
    const auto root_id = std::remove(moving_joint_ids.begin(),
                                     moving_joint_ids.end(), root_joint_idx);
    moving_joint_ids.erase(root_id, moving_joint_ids.end());
```
This way `pin.Model.getJointId()` is called only once